### PR TITLE
Tutorial

### PR DESCRIPTION
**Pull request by j3qq4hch**

### CODE REVIEW

**Comment by ctech4285** [comment]
wow I am glad I am not the only one submitting mergers titles 'blah'

**Comment by j3qq4hch** [comment]
I am doing that after you

**Comment by pavel-kirienko** [comment]
Blah. Sorry just felt like saying something.

**Review by pavel-kirienko — CHANGES_REQUESTED** [review]
Alexander, if you could replace the screenshots with English languaged ones, that would be great. Also please enable syntax highlighting, it works like that:

    ```c++
    int main()
    {
        std::puts("C++ code goes here");
        return 0;
    }
    ```

**Comment by pavel-kirienko** [comment]
You will likely greatly benefit from reading this https://guides.github.com/features/mastering-markdown/.

**Comment by j3qq4hch** [comment]
>>Alexander, if you could replace the screenshots with English languaged ones

I tried that, but I've got only Win10 ONE LANGUAGE version. That looks like nonsense, but I can't switch GUI language in my OS at all.

**Comment by pavel-kirienko** [comment]
This should help (crop as necessary):

virtualbox_win8_25_11_2016_08_33_30

You can also grab a free [Windows 10 image for VirtualBox from Microsoft directly (no joke)](https://www.microsoft.com/ru-RU/software-download/windows10).

**Comment by j3qq4hch** [comment]
Thanks!

**Comment by pavel-kirienko** [comment]
Hey @j3qq4hch, here's some feedback.

- [x] I propose to not spread the content over multiple articles, but join it instead into one large article. This would be coherent with the way the rest of the website is organized.
- [x] The code needs to be properly formatted and syntax highlighted (see my example above).
- [x] Tabs are not allowed anywhere, use spaces. Tabs do not render correctly.
- [x] Commands are case sensitive in some OS, so `Git` is not the same thing as `git`. I know that in Windows it wouldn't matter, but the universal convention is to always use lower case.
- [x] I believe we've discussed earlier that it makes sense to make different sections for Windows and Linux based hosts. The latter one would be empty for now, but it makes sense to structure the content in a way that allows for its easy extension later. E.g. you could put your "coming soon" under the Linux section as a space filler.
- [x] Please never forget to use proper typography. File system paths and code samples should be enclosed into a pair of single backticks (see the Markdown reference above). Also don't hesitate to explicitly structure the content using nested headers - this allows the website engine to generate neat and easily navigable table of contents on the left sidebar under the menu. For example, the two sections that describe how to flash the firmware via DCP and via the bootloader should definitely be nested into a common section.
- [x] Minor punctuation and stylistic errors, e.g. missing commas after "finally", and inappropriate upper case, e.g. BABEL instead of Babel. This is a very public website, so we must ensure that everything looks as perfect as possible. You could use the [user manual page for Babel](https://docs.zubax.com/zubax_babel) as an inspiration.
- [x] It's best to avoid C whenever possible. How about replacing all `THD_FUNCTION(....)` with `class MyThread : public chibios_rt::BaseStaticThread<512> { ... };`, and such? You can find plenty of examples in the main firmware. Compared to C, C++ offers higher expressiveness and better type safety, which allows to write more concise and reliable code. Also see [this](https://github.com/Zubax/zubax_style_guide).

**Comment by pavel-kirienko** [comment]
@j3qq4hch nudge

**Comment by j3qq4hch** [comment]
Indeed. Will fix it

**Comment by j3qq4hch** [comment]
You see, the main goal of this tutorial is to allow a person to start. I believe it is easier in C. After all, ChibiOS is written in C and its manuals and examples use C too.

**Comment by j3qq4hch** [comment]
BTW, why are we using english in this discussion?!

**Comment by pavel-kirienko** [comment]
Please check the page in a web browser, it is not rendering correctly.

The section dedicated to build dependencies should be split into Windows and Linux/OSX sub-sections.

Proper ordering of headings must be maintained. Heading level 1 is the name of the article, so only one such heading is allowed. Headers level 2+ are sections of the article. Heading level X may be a direct parent only to headers level X+1 (you are jumping from 1 to 3).

Typographic issues have not been fixed so I unchecked the corresponding item in the list above. You may find the single backtick syntax useful for highlighting one liners:

```
If everything is fine, you will find `compound.elf` in directory `/%reponame%/firmware/build/`.
```

I propose to provide complete source of `main.cpp` in the tutorials rather than diffs. Components that are known to not be used by the user application should also be removed, e.g. CAN and USB drivers. The diff based approach will leave the user with a ton of garbage code in their projects.

> You see, the main goal of this tutorial is to allow a person to start. I believe it is easier in C.

It would be a valid point if we were working on a general programming tutorial. Since we aren't, the learning curve argument should not apply. As for advantages of C++ over C in our case:

- C++ is probably a better choice for software development in general because of higher level abstractions and stronger safety guarantees. It makes sense to propagate better practices whenever possible.
- The main firmware is written in C++, so the reader will be able to use it as a reference.
- The Zubax ChibiOS libraries are written in C++ and have no C bindings. These libraries implement important features like non-volatile configuration parameter storage and some useful helpers.

Speaking of configuration parameters, why not include one or two in the second tutorial?

> why are we using english in this discussion?

This is a public repo, we don't know who could be reading.

**Comment by j3qq4hch** [comment]
>>The section dedicated to build dependencies should be split into Windows and Linux/OSX sub-sections.

There is no need to split this as it is absolutely the same on win and *nix
What really differs(not much though) is flashing the firmware. 

I must also say I'm not really good at C++ and much more used to C. 
Lets leave first blinking tutorials in C and the third one, which will use command line interface and parameter storing will be in C++.

**Comment by pavel-kirienko** [comment]
To further expand on this:

> I propose to provide complete source of main.cpp in the tutorials rather than diffs. Components that are known to not be used by the user application should also be removed, e.g. CAN and USB drivers. The diff based approach will leave the user with a ton of garbage code in their projects.

It might make sense to set up a dedicated directory for user-specific applications in the firmware repo as follows:

```
├── firmware
│   ├── ld.ld
│   ├── Makefile
│   ├── src/
│   └── zubax_chibios/
├── custom_firmware_template
│   ├── ld.ld                   - Symlink to firmware/ld.ld
│   ├── Makefile                - Custom
│   ├── src
│   │   ├── main.cpp            - Custom
│   │   ├── usb_cdc.hpp         - Copy from firmware/
│   │   ├── usb_cdc.cpp         - Sliglty modified copy from firmware/
│   │   ├── board/              - Sliglty modified copy from firmware/
│   │   └── sys/                - Sliglty modified copy from firmware/
│   └── zubax_chibios/          - Symlink to firmware/zubax_chibios
├── hardware
├── bootloader
└── <...>
```

**Comment by j3qq4hch** [comment]
I must admit it, WYSIWYG MD editors can't produce proper MD-pages for some reason. 
What do you use for preview?

**Comment by pavel-kirienko** [comment]
> There is no need to split this as it is absolutely the same on win and *nix

It isn't - gnu utils and Python need not be installed on anything but Windows.

> Lets leave first blinking tutorials in C and the third one, which will use command line interface and parameter storing will be in C++.

The only major difference between C and C++ in the first two tutorials would be the way threads are defined and launched. This is the correct way:

```c++
class MyThread : public chibios_rt::BaseStaticThread<1024>
{
    void main() override
    {
        setName("blinker");
        palSetPadMode(GPIOE, 8, PAL_MODE_OUTPUT_PUSHPULL);
        while (true)
        {
            palSetPad(GPIOE,8);
            chThdSleepMilliseconds(100);
            palClearPad(GPIOE, 8);
            chThdSleepMilliseconds(100);
        }
    }

public:
    virtual ~MyThread() { }
} thread_;

int main()
{
    auto watchdog = app::init();
    thread_.start(NORMALPRIO + 1);
    while (true)
    {
        watchdog.reset();
    }
}
```

**Comment by pavel-kirienko** [comment]
> I must admit it, WYSIWYG MD editors can't produce proper MD-pages for some reason.
> What do you use for preview?

I write in a regular text editor and use a local instance of the website for preview.

**Comment by pavel-kirienko** [comment]
In the second tutorial, `pwmcfg` must be declared `const`.

**Comment by pavel-kirienko** [comment]
Nudge nudge

**Comment by ctech4285** [comment]
This is too long for me to read

**Comment by pavel-kirienko** [comment]
We'll make a movie out of it someday.

**Comment by ctech4285** [comment]
We shall call it Moscow Valley

**Comment by pavel-kirienko** [comment]
@j3qq4hch You didn't review the changes, apparently. Please always do so.

C sections are not needed, we should constrain ourselves to C++ only.

File: 0 Products/40 Zubax Babel/Tutorials/10 Developing custom applications.md [hunk 1]
```diff
@@ -1,4 +1,263 @@
...
+    while (true) 
+    {
+        palSetPad(GPIOE,8);
+        chThdSleepMilliseconds(100);
+        palClearPad(GPIOE, 8);
+        chThdSleepMilliseconds(100);
+    }
+}
+```
+Now find function ```int main()``` in ```zubax_babel\firmware\src\main.cpp```. This is program entry point. Comment it all and write your own version like this:
```
Comment (pavel-kirienko) [inline 0 Products/40 Zubax Babel/Tutorials/10 Developing custom applications.md]: Here and elsewhere: inline code blocks are highlighted with single backticks: ``` `code` ```. Triple backticks are needed to highlight multi-line code blocks.

File: 0 Products/40 Zubax Babel/Tutorials/10 Developing custom applications.md [hunk 2]
```diff
@@ -1,4 +1,263 @@
...
+- [Python 3+](https://www.python.org/downloads/)
+- gnu arm toolchain. It is very version-dependant, so you should use [this version](https://launchpad.net/gcc-arm-embedded/4.9/4.9-2015-q3-update)
 
+After you download and install everything make sure you have all these your environment variable PATH set up correctly and have access to all necessary utilities from console. For example, my PATH variable looks like this:
+
+<img src="path.png" class="thumbnail" title="PATH">
+
+Now its time to try to build the original firmware. Go to console and clone the repository
+
+```git clone https://github.com/Zubax/zubax_babel```
```
Comment (pavel-kirienko) [inline 0 Products/40 Zubax Babel/Tutorials/10 Developing custom applications.md]: Are you sure it is necessary to put such a simple command on a separate line? Also, missing punctuation at the end of the previous line.

File: 0 Products/40 Zubax Babel/Tutorials/10 Developing custom applications.md [hunk 3]
```diff
@@ -1,4 +1,263 @@
...
+### BREATH (in C) ###
+Now we have blinking led, but I find it a little bit rough. Let’s make it glow. We will need PWM. According to babel schematic one of the LEDS is connected to channel4 of timer3.
+In order to enable timer3 pwm generation ChibiOS must be configured a little bit. 
+
+
+Find macro ```HAL_USE_PWM``` in ```halconf.h``` and make it `TRUE`
+
+Find macro `STM32_PWM_USE_TIM3` in ```mcuconf.h``` and make it `TRUE`
+
+Now add breath_thread to you code:
```
Comment (pavel-kirienko) [inline 0 Products/40 Zubax Babel/Tutorials/10 Developing custom applications.md]: Broken formatting

File: 0 Products/40 Zubax Babel/Tutorials/10 Developing custom applications.md [hunk 4]
```diff
@@ -1,4 +1,263 @@
...
+
+Load it and watch green led blinking ON and OFF and red led smoothly chaging its brightness. Isn’t it fancy?
+
+---
+
+### BLINK and BREATH in C++ ###
+Now when we can blink a LED differently lets make it in a more civilized way - in C++
+Although Chibios is written in C it offers rich C++ capabilities and we can rewrite out blinking threads like this:
+
+```
```
Comment (pavel-kirienko) [inline 0 Products/40 Zubax Babel/Tutorials/10 Developing custom applications.md]: Missing syntax highlight directive: `c++`

**Comment by j3qq4hch** [comment]
Come on. Let's leave a couple of them just to demonstrate a variety of possibilities.

File: 0 Products/40 Zubax Babel/Tutorials/10 Developing custom applications.md [hunk 5]
```diff
@@ -1,4 +1,263 @@
...
+    virtual ~BlinkerThread() { }
+} blinker_thread_;
+```
+
+and
+
+```
+static const PWMConfig pwmcfg = {
+  10000,                                    //10KHz PWM clock frequency.   
+  255,                                      //255 ticks is pwm resolution
```
Comment (pavel-kirienko) [inline 0 Products/40 Zubax Babel/Tutorials/10 Developing custom applications.md]: Missing space at the beginning of the comments. I admit our style guide does not say anything about this, this is my fault; I will update it right now.

**Review by pavel-kirienko — CHANGES_REQUESTED** [review]

File: 0 Products/40 Zubax Babel/Tutorials/10 Developing custom applications.md (hunk 4 above)
Comment (j3qq4hch) [reply 0 Products/40 Zubax Babel/Tutorials/10 Developing custom applications.md]: So we need to highlight programming language name too?

**Comment by pavel-kirienko** [comment]
> variety of possibilities

What for? C is generally inferior to C++ in terms of code readability, safety, and performance (when used properly) (I hate both though). Keeping C code samples would only increase maintenance workload.

**Comment by j3qq4hch** [comment]
What if some c-guy(like me) would like to use babel?

**Comment by pavel-kirienko** [comment]
They would have to learn basic C++, it's not hard, and there are **helpful tutorials** to the rescue!

**Comment by j3qq4hch** [comment]
I give up. I will remove C-sections

**Comment by pavel-kirienko** [comment]
Embrace C++ Alexander.

image

**Comment by j3qq4hch** [comment]
Sounds like 
-Hey, mister, do you have a minute to discuss our holy god Jesus Christ?

**Comment by pavel-kirienko** [comment]
Don't blow my cover please.

**Comment by pavel-kirienko** [comment]
Is it ready for review Alexander?

**Comment by j3qq4hch** [comment]
I'm about to add some more today. Wait a bit

**Comment by pavel-kirienko** [comment]
@j3qq4hch if you have anything worth saving in your branch, please save it locally, then delete it.

**Pull request closed without merging**